docs/general/developer-resources/emerald-paratime/writing-dapps-on-emerald.md

```diff
@@ -614,6 +614,434 @@ share them with us at [#dev-support on our Community Slack channel][slack-devsup
+hex = "0.4.3"
+tokio = { version = "1.17", features = [ "rt", "macros" ] }
+```
+
+The following `main.rs` will compile the solidity smart contract by calling `solc` in the backgroung.
+After that, it saves the ABI to a file so that it can be used later in order to call the contract.
+There is no need to save the bytecode to disk, as you can deploy it directly to Emerald.
+
+The script is expecting two environment variables: `EMERALD_IP` and `PRIVATE_KEY`. In case you don't
+specify `EMERALD_IP` it will default to `127.0.0.1` just in case you are
+[running a local server](#running-private-oasis-network-locally). The private key must be specified
+in order to get be able to deploy the contract.
+
+```ts
+use ethers::{contract::ContractFactory, prelude::*, solc::Solc};
+use eyre::Result;
+use std::{convert::TryFrom, fs::File, io::Write, sync::Arc, time::Duration};
+
+#[tokio::main]
+async fn main() -> Result<()> {
+    // Load contract and compile. This requires the solc binary.
+    // In case your contract does not compile, check the errors with solc.
+    let contract_name = "Greeter";
+    let contract = "../solidity/Greeter.sol";
+    let contracts = Solc::default().compile_source(&contract)?;
+    let c = contracts.get(&contract, &contract_name).unwrap();
+    let abi = c.abi.unwrap().to_owned();
+    let bytecode = c.bytecode().unwrap().to_owned();
+
+    // save the ABI to a file in order to load it in the call script
+    let abi_json = serde_json::to_string_pretty(&abi)?;
+    let abi_json_file_path = "/path/to/Greeter.abi";
+    let mut abi_json_file_handler = File::create(abi_json_file_path)?;
+    write!(abi_json_file_handler, "{}", abi_json)?;
+
+    // set up emerald-related variables (blockchain url, wallet, client)
+    let emerald_ip = std::env::var("EMERALD_IP").unwrap_or_else(|_| "127.0.0.1".into());
+    let emerald_uri = format!("http://{}:8545", emerald_ip);
+    let private_key = std::env::var("PRIVATE_KEY").expect("PRIVATE_KEY env var is mandatory");
+    let private_key = private_key.trim_start_matches("0x");
+    let provider = Provider::<Http>::try_from(emerald_uri)?.interval(Duration::from_secs(7));
+    let chain_id = provider.get_chainid().await?.as_u64();
+    let wallet = private_key.parse::<LocalWallet>()?.with_chain_id(chain_id);
+    println!("wallet public address is {:?}", wallet.address());
+    let client = Arc::new(SignerMiddleware::new(provider, wallet));
+
+    // deploy the contract
+    let factory = ContractFactory::new(abi, bytecode, client);
+    let contract = factory
+        .deploy("HELLO FROM RUST!".to_string())?
+        .legacy()
+        .confirmations(0usize)
+        .send()
+        .await?;
+    println!("Contract address: {:?}", contract.address());
+
+    Ok(())
+}
+```
+
+In case you don't want to trigger `solc` from rust, you can also compile the bytecode directly
+from bash by running `solc --abi --bin -o . Greeter.sol` and after that loading the ABI and the bytecode
+from the files:
```

Comment:
Yes, invoking `solc` from Rust is a bad idea. See my comment above and adapt the `cargo build` rule.

Reply:
This would unify the build phase for Rust and Go and would simplify the entire article. I'll adapt it based on your suggestions.